### ISSUE TITLE
Protobuf repo recompilation sensitivity

### ISSUE THREAD

**Issue by pcj** [bazelbuild/bazel#7095]
As discussed on the mailing list https://groups.google.com/forum/#!topic/bazel-discuss/UKLK7Y-jsW4 the protobuf repo is notoriously sensitive to recompilation.  Presumably something in the environment used in the action key for protobuf C++ compilation leads to cache invalidation and hence recompiling protobuf from source.  

Given protobuf libs are often base dependencies for many others, leads to degraded user experience.

Repro steps include taking any workspace that has a proto dependency and `source ~/.bashrc`, watch protobuf recompile.

Observed up to and including bazel 0.20.0.

/assign @hlopko

**Comment by hlopko** [comment]
Can you double check that:

	git clone https://github.com/bazelbuild/rules_cc
	cd rules_cc
	bazel test //tools/migration/...
	source ~/.bashrc
	bazel test //tools/migration/...

results in recompilation? What is the minimal content of the ~/.bashrc file that triggers the issue? What platform are you on?

**Comment by pcj** [comment]
Looks like it's the `$PATH`.  Any changes to $PATH will trigger recompilation.  

Perhaps this is working as intended, or perhaps we can split it and use a set to dedup entries to normalize the path to the minimal necessary set.  

Is it possible to exclude `PATH` here and rely solely on the toolchain?

**Comment by pcj** [comment]
```sh
$ uname -a
Linux bina 4.13.0-46-generic #51-Ubuntu SMP Tue Jun 12 12:36:29 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by pcj** [comment]
Yes!  Tested on bazel 0.21.0 and indeed no longer recompiles with `PATH` change.  I guess I will go ahead and close this.  The migration path from 0.19.2 to 0.20.0 and 0.21.0 is a little challenging (for this same reason), but clearly beneficial.  Thanks for your help @hlopko.

**Comment by hlopko** [comment]
Awesome, thanks Paul. Can you pls share this solution on the mailing list?

**Comment by hlopko** [comment]
@pcj What toolchain do you use? Was it the Bazel's autoconfigured C++ toolchain?

**Comment by hlopko** [comment]
Also, on what platforms do you observe this?

**Comment by pcj** [comment]
@hlopko I had observed it on linux & darwin before bazel 0.21.0.  Now that I'm on 0.21 I'm really not seeing this anymore. 

For most projects I'm not using a special toolchain, so yes, the autoconfigured one.

**Comment by buchgr** [comment]
I believe that I found the root cause. It looks like this issue is not specific to C++, but that almost all actions (spawn actions, c++ compile, c++ link, java compile) in Bazel declare a dependency on `PATH`  independent of whether these actions actually use them. It's probably just most obvious with c++ actions as they typically take quite long.

**Comment by keith** [comment]
That would make sense with our solution to this which was to launch bazel with `env -i` and only pass very limited variables (and a static PATH variable)

**Comment by meisterT** [comment]
@buchgr friendly ping

**Comment by buchgr** [comment]
@meisterT what's up? :)

**Comment by meisterT** [comment]
I have heard from bazel users that this is still a pain point. So I wanted to check in if you are working on a fix and when we can expect it to land?

**Comment by buchgr** [comment]
Yes I am working on a proposal to fix it. I don't want to make promises on when it will land. 1-2 quarters.

**Comment by sudoforge** [comment]
I currently suffer from the protobuf targets rebuilding (a dependency of `bazelbuild/buildtools` (specifically, buildifier), but cannot reproduce the issue by re-sourcing my shell's rc file. To be honest, I'm really not sure how to debug this further other than "it happens every 5-10 minutes of inactivity".

**Comment by Anduril-Xu** [comment]
Its been annoying for a while.

In my setting, I have several languages under the same workspace, python, java, javascript/typescript. I use Intellij for most of the development, but every time I build one of the targets in one of the languages, the bazel cache seems to be invalidated for other languages, thus triggering a rebuild of some packages, most notably protobuf. It is only 10-20 seconds each time, but very annoying when I do cross language development.

Is this expected? anyway to not rebuild protobuf?

**Comment by franciscocpg** [comment]
> Its been annoying for a while.
> 
> In my setting, I have several languages under the same workspace, python, java, javascript/typescript. I use Intellij for most of the development, but every time I build one of the targets in one of the languages, the bazel cache seems to be invalidated for other languages, thus triggering a rebuild of some packages, most notably protobuf. It is only 10-20 seconds each time, but very annoying when I do cross language development.
> 
> Is this expected? anyway to not rebuild protobuf?
> 
> EDIT:
> my bazel version was 0.29.1. the latest 3.2.0 also does not solve the problem
> 
> EDIT again:
> seems like using prebuilt binaries works, e.g. https://github.com/google/startup-os/blob/master/WORKSPACE#L142-L154

Hi @Anduril-Xu 
Were you able to configure your workspace to use prebuilt binaries? I have a go monorepo and haven't had success yet :disappointed:

**Comment by sudoforge** [comment]
@franciscocpg Are you still experiencing this? Have you tried using `--incompatible_strict_action_env`?

**Comment by franciscocpg** [comment]
Hi @sudoforge 

I want to use prebuilt binaries and have asked about it here: https://github.com/bazelbuild/rules_go/issues/2684.

But I haven't tried `http_file` yet. Will give it a try.

**Comment by c-mita** [comment]
@Anduril-Xu - I tried replicating recompilation of protobuf when changing language targets. The only issue I saw seemed to have been caused by the relatively recent change (#12430) that was made after your comment. (And my merged PR on the protobuf repo addresses it (https://github.com/protocolbuffers/protobuf/pull/8626)).

Is this still an issue?

**Comment by stevekuznetsov** [comment]
This is definitely still an issue - to be honest in projects that use Bazel I've never actually seen it "fixed" - I just got used to the fact that protobuf recompiles all the time. Hit this just today in `cockroachdb/cockroach` with 

```
$ bazel version
Bazelisk version: v1.11.0
Build label: 5.1.0
Build target: bazel-out/k8-opt/bin/src/main/java/com/google/devtools/build/lib/bazel/BazelServer_deploy.jar
Build time: Thu Mar 24 14:02:30 2022 (1648130550)
Build timestamp: 1648130550
Build timestamp as int: 1648130550
```

**Comment by stevekuznetsov** [comment]
Seems like running gazelle can invalidate and cause a recompile for this package?

**Comment by meisterT** [comment]
@stevekuznetsov Can you write up repro instructions for cockroachdb w. gazelle please?

**Comment by stevekuznetsov** [comment]
@meisterT pinning down a direct reproducer is tricky. Cross-linked to an issue on the Cockroach project where devs talk about other targets getting recompiled, with some theories about why.

**Comment by aaliddell** [comment]
I've tried to trace this cache fragility before and thought it might be https://github.com/protocolbuffers/protobuf/issues/6886. However, IIRC that didn't solve it. When I've triggered this cache issue before with execution logging turned on, the action hashes between builds were identical but it was still rebuilding, suggesting it is not exclusively an env related issue.

One consideration may be that protobuf is treated somewhat special by Bazel, in that there are native flags and rules that reference it. As such, this may be why protobuf behaves unlike any other external repo with regard to caching. Since the execution logs suggest there should be no rebuild, is there a way of tracing deeper within Bazel as to why it made a decision to rebuild?

**Comment by pcj** [comment]
Presumably this would all go away if the proto rules were starlarkified as has been discussed for several years.  Any update from the bazel team on this?

**Comment by meisterT** [comment]
cc @comius

**Comment by dt** [comment]
I spent a little time this morning looking into why I was seeing very frequent protoc recompilations in cockroachdb/cockroach development and I believe that in my case might also be down to PATH changes as described above, but what was surprising to me is that this is a result of using the rules_go recommended editor configuration for my editor (VS Code). 

My editor is configured using [the recommended configuration](https://github.com/bazelbuild/rules_go/wiki/Editor-setup#visual-studio-code), which includes both a) pointing gopls at bazel and gopackagesdriver and b) setting `goroot` to `${workspaceFolder}/bazel-${workspaceFolderBasename}/external/go_sdk`. 

Configuring the goroot seems more correct than letting vscode go invoke system go, so that when it runs `go mod tidy` or whatever, it is using the exact same version and installation of the sdk as bazel is when building. It appear to do so by prepending the configured path to PATH when invoking all tools. In isolation, this seems good/correct. 

However this mean that when gopls is invoking bazel and gopackagesdriver, it is doing so with this modified PATH variable. Thus bazel invocations in my workspace -- between my direct ones and the gopls ones -- are constantly switching between this modified PATH and my usual PATH.

Simulating this thrashing by hand by running `PATH=_bazel/path/to/sdk:$PATH bazel build some/tiny/pkg` and then a plain `bazel build some/tiny/pkg` and back again produces the same effect, as expected, of invalidating protoc.

I guess there's no new information here as far as protoc's sensitivity to PATH changes, but it wasn't obvious to me going in here that my PATH was changing so often as a result of using the recommended configuration.

**Comment by sudoforge** [comment]
> it wasn't obvious to me going in here that my PATH was changing so often as a result of using the recommended configuration.

yeah, this is really a bit of hidden knowledge that you learn through trial like you, but due to the sensitivity of `PATH` changes here, it's best to always execute `bazel` in the same environment. i've come to the conclusion that this means either focusing on

A) only executing bazel in the same environment locally, e.g. always via VS Code, or always via your shell; or
B) setting up RCE and always relying on that

**Comment by dt** [comment]
For what it's worth, the executions "via" vs code here aren't voluntary ie me triggering builds in my editor versus from my terminal -- they are the result of the recommended editor configuration which changes its code navigation/inspection (via gopls) to be backed by bazel and gopackagedriver (which seems good!). These invocations happen "out of sight" and in the background so it wasn't obvious to me, at first, that they were what was thrashing PATH.

**Comment by pdeva** [comment]
i dont thonk the vscode config is the rooti ssue. we use vscode too, but point it to manually installed Go. so no changes to PATH or anything. yet merely afterbrunning gazelle to update a minor dependency rebuilds the entire protobuf stuff and thus the entire rest of the repo for some reason.

**Comment by fmeum** [comment]
@pdeva If you can minimize your repo to a reproducer you can share, I can investigate the rebuilds.

**Comment by dt** [comment]
Just one last update from my digging, just in case it can save anyone else some time if they're in a similar situation:

I dug a bit more and realized that in our particular repo (cockroachdb/cockorach), we had developed particular sensitivity to PATH changes earlier this year when a developer added `action_env=PATH`  to our dev config. It looks like this was motivated by new arm-based Macs having a different default location for homebrow-installed tools, including those needed by foreign_cc, now under /opt instead of /usr/local, only the latter of which is in the  hardcoded default used by foreign_cc. Using `action_env=PATH` to let the user's existing configuration of platform-specific paths probably seemed correct/preferred at the time and in isolation, given we didn't realize tools would be invoking bazel with customized PATHs.

I have since changed our common dev configuration back to using a hardcoded PATH, based the default from foriegn_cc just with the addition of the new /opt homebrew location, and so have not heard any additional reports of "spurious" protoc rebuilds 🤞.

Again, this one is probably pretty obvious in hindsight -- having a tool config that changed the PATH constantly and a configuration that makes compilation depend on PATH means constant recompilation -- but in isolation each of the steps involved to arrive at that situation seemed innocuous enough at the time. I just wanted to spell it out on the off chance it saves anyone else a little debugging time.

**Comment by nioan-aux** [comment]
I'm wondering if there are any updates on this. It's affecting our build as well. 

Would it be possible to implement a work around to expose a flag that allows for a pre-build version of grpc to be used instead of rebuilding it. Pantsbuild is using this approach.

This has been open for 5+ years, and it seems quite tricky to reproduce and find the root cause for.  A workaround might be a good interim solution

**Comment by fzakaria** [comment]
For those curious, here are two compact execution logs.
All I did was change `PATH` before running Bazel.

```
> bazel-bin/src/tools/execlog/parser \
  --log_path=exec_protoc1.log \
  --log_path=exec_protoc2.log \
  --output_path=/tmp/exec1.log.txt \
  --output_path=/tmp/exec2.log.txt
```

I looked at the output.

```
> diff /tmp/exec1.log.txt /tmp/exec2.log.txt | head
67c67
<   value: "/Users/fzakaria/Library/Caches/bazelisk/downloads/sha256/7dc88df4842a5b49830c55b174551445587ae60f027a67d0f847dbb35f1f40c6/bin:/Users/fzakaria/.pyenv/shims:/Users/fzakaria/.nix-profile/bin:/nix/var/nix/profiles/default/bin:/usr/local/bin:/System/Cryptexes/App/usr/bin:/usr/bin:/bin:/usr/sbin:/sbin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/local/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/appleinternal/bin:/opt/homebrew/bin:/Applications/iTerm.app/Contents/Resources/utilities:/Users/fzakaria/code/github.com/confluentinc/cc-dotfiles/bin/access-ga:/Users/fzakaria/code/github.com/confluentinc/cc-dotfiles/bin/cloud-ga:/Users/fzakaria/code/github.com/confluentinc/cc-dotfiles/bin/connect-ga:/Users/fzakaria/code/github.com/confluentinc/cc-dotfiles/bin/k8saas-ga:/Users/fzakaria/code/github.com/confluentinc/cc-dotfiles/bin/devprod-ga:/Users/fzakaria/code/github.com/confluentinc/cc-dotfiles/bin/traffic-ga:/Users/fzakaria/code/github.com/confluentinc/cc-dotfiles/bin/ksql-ga:/Users/fzakaria/code/github.com/confluentinc/cc-dotfiles/bin/orchestrator-ga:/Users/fzakaria/code/github.com/confluentinc/cc-dotfiles/bin/kafka-ga:/Users/fzakaria/code/github.com/confluentinc/cc-dotfiles/bin/servicefoundations-beta:/Users/fzakaria/.zsh/plugins/powerlevel10k"
---
>   value: "/Users/fzakaria/Library/Caches/bazelisk/downloads/sha256/7dc88df4842a5b49830c55b174551445587ae60f027a67d0f847dbb35f1f40c6/bin:/Users/fzakaria/.pyenv/shims:/Users/fzakaria/.nix-profile/bin:/nix/var/nix/profiles/default/bin:/usr/local/bin:/System/Cryptexes/App/usr/bin:/usr/bin:/bin:/usr/sbin:/sbin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/local/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/appleinternal/bin:/opt/homebrew/bin:/Applications/iTerm.app/Contents/Resources/utilities:/Users/fzakaria/code/github.com/confluentinc/cc-dotfiles/bin/access-ga:/Users/fzakaria/code/github.com/confluentinc/cc-dotfiles/bin/cloud-ga:/Users/fzakaria/code/github.com/confluentinc/cc-dotfiles/bin/connect-ga:/Users/fzakaria/code/github.com/confluentinc/cc-dotfiles/bin/k8saas-ga:/Users/fzakaria/code/github.com/confluentinc/cc-dotfiles/bin/devprod-ga:/Users/fzakaria/code/github.com/confluentinc/cc-dotfiles/bin/traffic-ga:/Users/fzakaria/code/github.com/confluentinc/cc-dotfiles/bin/ksql-ga:/Users/fzakaria/code/github.com/confluentinc/cc-dotfiles/bin/orchestrator-ga:/Users/fzakaria/code/github.com/confluentinc/cc-dotfiles/bin/kafka-ga:/Users/fzakaria/code/github.com/confluentinc/cc-dotfiles/bin/servicefoundations-beta:/Users/fzakaria/.zsh/plugins/powerlevel10k:/bin2/"
568,569c568
< runner: "disk cache hit"
< cache_hit: true
---
> runner: "darwin-sandbox"
573c572
```

The environment variable is part of the action.
Is there some special passing of PATH for CC actions?

Is there a reason why PATH would be part of the action itself and not the digest of the tools itself for instance.

[exec_protoc1.log](https://github.com/user-attachments/files/17498498/exec_protoc1.log)
[exec_protoc2.log](https://github.com/user-attachments/files/17498499/exec_protoc2.log)

**Comment by fzakaria** [comment]
So we can probably close this then?

**Comment by fmeum** [comment]
I would say yes as there is nothing about this that's specific to protobuf. It would be great to continue the discussion about why `PATH` is set on C++ compilation actions in a more narrowly scoped, new issue.

**Comment by DeCarabas** [comment]
I will say that when I was investigating and debugging this I had a hermetic toolchain. Every compiler binary I needed was fetched by bazel and I provided the exact path to each tool as required.

I understand needing PATH when you're using a system compiler but... 

(I agree that this probably isn't directly related to protobuf but please carry this forward into future discussions...)

**Comment by fzakaria** [comment]
@fmeum @DeCarabas -- A correct approach I've been thinking about is for the ActionKey to use the digest of the tools it uses and not the environment variables;

So if you find `clang` from PATH, I don't necessarily care about how it was found, but instead that it's the **exact same** compiler. (Might need to extend this to the whole transitive graph of the binary with `ldd`)

This feels more systemic to how Bazel tracks system dependencies that can make reliance of the system a lot simpler.
That way if two machines are deployed with the same `clang` binary installation, they can get same caching behavior?

**Comment by nioan-aux** [comment]
As a a workraround, moving `/usr/local/bin/bazel` to `/usr/local/bin/bazel-raw` and then adding the following script in `/usr/bin/local/bazel` is working great for us.  The --action_env arg was not working as good as the analysis was invalided when changing the PATH (e.g. when changing conda envs). 

```
# Add any other paths you need for host tools
export PATH=/usr/bin:/usr/local/bin

/usr/local/bin/bazel-raw "$@"
```

**Comment by fzakaria** [comment]
@nioan-aux if you didn't know this; you can put a file `//tools/bazel` in your repository and Bazel will automatically invoke that as a wrapper.

**Comment by jguamie** [comment]
[Pre-built binary toolchain for protoc](https://github.com/protocolbuffers/protobuf/issues/19558) has been prioritized and should be completed by EOY 2025.

**Comment by fmeum** [comment]
Prebuilt protoc toolchains have landed in v33.4 (https://github.com/protocolbuffers/protobuf/issues/19558#issuecomment-3756666746). Usage is explained in https://github.com/protocolbuffers/protobuf/issues/19558#issuecomment-3760300140, where the incompatible flag will be enabled by default in Bazel 9.
